src/geom/prep/AbstractPreparedPolygonContains.cpp

```diff
@@ -109,23 +109,38 @@ AbstractPreparedPolygonContains::eval(const geom::Geometry* geom)
 {
     // Do point-in-poly tests first, since they are cheaper and may result
     // in a quick negative result.
-    //
-    // If a point of any test components does not lie in target,
-    // result is false
-    bool isAllInTargetArea = isAllTestComponentsInTarget(geom);
-    if(!isAllInTargetArea) {
+    auto outermostLoc = getOutermostTestComponentLocation(geom);
+
+    if (outermostLoc == Location::EXTERIOR) {
+        // If a point of any test components does not lie in target,
+        // result is false
         return false;
     }
 
-    // If the test geometry consists of only Points,
-    // then it is now sufficient to test if any of those
-    // points lie in the interior of the target geometry.
+    // Short-circuit for (Multi)Points
+    //
+    // For the Covers predicate, we can return true
+    // since no Points are on the exterior of the Polygon.
+    //
+    // For the Contains predicate, we need to test if any
+    // of those points lie in the interior of the target
+    // geometry.
+    //
     // If so, the test is contained.
     // If not, all points are on the boundary of the area,
     // which implies not contained.
-    if(requireSomePointInInterior && geom->getDimension() == 0) {
-        bool isAnyInTargetInterior = isAnyTestComponentInTargetInterior(geom);
-        return isAnyInTargetInterior;
+    if (geom->getDimension() == 0) {
```

Comment:
How about this code?  It puts the short-circuits first, to reduce depth of IF blocks.  Also names the test booleans rather than having them buried in IF conditions.  Easier to understand, I think.

```
    if (geom->getDimension() == 0) {
       // covers evaluation
        if (! requireSomePointInInterior)  return true;

       // contains evaluation
        boolean hasInteriorPoint = false;
        if (geom->getNumGeometries() == 1) {
                // Only  one point
                hasInteriorPoint = outermostLoc == Location::INTERIOR;
            }
        else {
           // for MultiPoint, check if there is at least one point in interior
            hasInteriorPoint = isAnyTestComponentInTargetInterior(geom);
       }
       return hasInteriorPoint;
    }
```

Reply:
(code should be tested... I think it's right, but inverting all the logic is definitely a bit error-prone)

Reply:
It might be even clearer to refactor this logic into a separate internal method, which is focussed on just handling points.  This makes the short-circuit for points very obvious.

Like this:

```
if (geom->getDimension() == 0) {
  return evalPointTarget(geom, outermostLoc);
}

boolean evalPointTarget(Geometry geom, Location outermostLoc)
{
       // covers evaluation
        if (! requireSomePointInInterior)  return true;

       // contains evaluation
        boolean hasInteriorPoint = false;
        if (geom->getNumGeometries() == 1) {
                // Only  one point
                hasInteriorPoint = outermostLoc == Location::INTERIOR;
            }
        else {
           // for MultiPoint, check if there is at least one point in interior
            hasInteriorPoint = isAnyTestComponentInTargetInterior(geom);
       }
       return hasInteriorPoint;
}
```

Reply:
I think this formulation subjects MultiPoints to a redundant PIP test if `outermostLoc == Location::INTERIOR`.

Reply:
Refactored in 51c2b64; hopefully this captures what you're after without adding the extra test.

Reply:
Great.  Since the code doesn't make it fully clear that there is an ordering relation on Location, it would be better to define this explicitly in the Location class, and then use that here.  Perhaps a method `Location::Index`?  That would allow a simple comparison to be used in this method.

Reply:
It would also be nice to run all the predicate unit tests through the Prepared code.  Not sure if there's a portable way to do this.  A quick-n-dirty way is just to hack the XMLTester code for each predicate to run both.